src/base/ClpMap.h

```diff
@@ -0,0 +1,284 @@
+    uint64_t memLimit_ = 0;
+
+    /// the total amount of memory we currently use for all cached entries
+    uint64_t memUsed_ = 0;
+};
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+ClpMap<Key, Value, MemoryUsedBy>::ClpMap(const uint64_t capacity, const Ttl defaultTtl):
+    defaultTtl_(defaultTtl)
+{
+    assert(defaultTtl >= 0);
+    setMemLimit(capacity);
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+void
+ClpMap<Key, Value, MemoryUsedBy>::setMemLimit(const uint64_t newLimit)
+{
+    assert(newLimit >= 0);
+    if (memUsed_ > newLimit)
+        trim(memLimit_ - newLimit);
+    memLimit_ = newLimit;
+}
+
+/// \returns the index position of an entry identified by its key (or end())
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+typename ClpMap<Key, Value, MemoryUsedBy>::IndexIterator
+ClpMap<Key, Value, MemoryUsedBy>::find(const Key &key)
+{
+    const auto i = index_.find(key);
+    if (i == index_.end())
+        return i;
+
+    const auto entryPosition = i->second;
+    if (!entryPosition->expired()) {
+        if (entryPosition != entries_.begin())
+            entries_.splice(entries_.begin(), entries_, entryPosition);
+        return i;
+    }
+    // else fall through to cleanup
+
+    erase(i);
+    return index_.end();
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+const Value *
+ClpMap<Key, Value, MemoryUsedBy>::get(const Key &key)
+{
+    const auto i = find(key);
+    if (i != index_.end()) {
+        const auto &entry = *(i->second);
+        return &entry.value;
+    }
+    return nullptr;
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+Optional<uint64_t>
+ClpMap<Key, Value, MemoryUsedBy>::MemoryCountedFor(const Key &k, const Value &v)
+{
+    // Both storage and index store keys, but we count keySz once, assuming that
+    // copying a Key does not consume more memory. This assumption holds for
+    // Key=SBuf, but, ideally, we should be outsourcing this decision to another
+    // configurable function, storing each key once, or hard-coding Key=SBuf.
+    const auto keySz = k.length();
+
+    // approximate calculation (e.g., containers store wrappers not value_types)
+    return SafeSum<uint64_t>(
+        keySz,
+        // storage
+        sizeof(typename Entries::value_type),
+        MemoryUsedBy(v),
+        // index
+        sizeof(typename Index::value_type));
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+bool
+ClpMap<Key, Value, MemoryUsedBy>::add(const Key &key, const Value &v, const Ttl ttl)
+{
+    // optimization: avoid del() search, MemoryCountedFor() in always-empty maps
+    if (memLimit() == 0)
+        return false;
+
+    del(key);
+
+    if (ttl < 0)
+        return false; // already expired; will never be returned by get()
+
+    const auto memoryRequirements = MemoryCountedFor(key, v);
```

Comment:
This add() method is the only place where we use MemoryUsedBy template parameter today. The code that calls add() probably knows what entry it is adding, so it could supply the sizing information to add(). The current MemoryUsedBy() default value is unused IIRC. If we ever add support for dynamic size changes this single sizing function is not going to be enough anyway. Should we remove the MemoryUsedBy parameter and add the corresponding size parameter instead?

Reply:
Tempting, but many of the caches that will be migrated to using this will only need the default calculation. So for now I will leave it in and we can address the differences(if any) for dynamic entries when that happens.